LogStreamBatchWriterTest.shouldWriteEventWithMetadataBuffer is flaky

**Describe the bug**
Fails sometimes with:

```
[ERROR] Tests run: 15, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 9.9 s <<< FAILURE! - in io.zeebe.logstreams.log.LogStreamBatchWriterTest
[ERROR] shouldWriteEventWithMetadataBuffer(io.zeebe.logstreams.log.LogStreamBatchWriterTest)  Time elapsed: 0.163 s  <<< FAILURE!
java.lang.AssertionError: 

Expecting:
 <-1L>
to be greater than:
 <0L> 
	at io.zeebe.logstreams.log.LogStreamBatchWriterTest.getWrittenEvents(LogStreamBatchWriterTest.java:68)
	at io.zeebe.logstreams.log.LogStreamBatchWriterTest.shouldWriteEventWithMetadataBuffer(LogStreamBatchWriterTest.java:226)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.rules.ExpectedException$ExpectedExceptionStatement.evaluate(ExpectedException.java:239)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:55)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:137)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeLazy(JUnitCoreWrapper.java:119)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:87)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:158)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
```

--- a/logstreams/src/test/java/io/zeebe/logstreams/log/LogStreamBatchWriterTest.java
+++ b/logstreams/src/test/java/io/zeebe/logstreams/log/LogStreamBatchWriterTest.java
@@ -15,6 +15,7 @@ import io.zeebe.logstreams.impl.LoggedEventImpl;
 import io.zeebe.logstreams.util.LogStreamReaderRule;
 import io.zeebe.logstreams.util.LogStreamRule;
 import io.zeebe.logstreams.util.LogStreamWriterRule;
+import io.zeebe.test.util.TestUtil;
 import io.zeebe.util.buffer.BufferUtil;
 import io.zeebe.util.buffer.DirectBufferWriter;
 import io.zeebe.util.sched.future.ActorFuture;
@@ -22,6 +23,7 @@ import io.zeebe.util.sched.testing.ControlledActorSchedulerRule;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.concurrent.ExecutionException;
+import java.util.function.Consumer;
 import org.agrona.DirectBuffer;
 import org.agrona.concurrent.UnsafeBuffer;
 import org.junit.Before;
@@ -106,10 +108,15 @@ public class LogStreamBatchWriterTest {
     return new UnsafeBuffer(buffer, offset, length);
   }
 
+  private long write(Consumer<LogStreamBatchWriter> consumer) {
+    consumer.accept(writer);
+    return TestUtil.doRepeatedly(() -> writer.tryWrite()).until(pos -> pos > 0);
+  }
+
   @Test
   public void shouldReturnPositionOfSingleEvent() {
     // when
-    final long position = writer.event().keyNull().value(EVENT_VALUE_1).done().tryWrite();
+    final long position = write(w -> w.event().keyNull().value(EVENT_VALUE_1).done());
 
     // then
     assertThat(position).isGreaterThan(0);
@@ -123,16 +130,16 @@ public class LogStreamBatchWriterTest {
   public void shouldReturnPositionOfLastEvent() {
     // when
     final long position =
-        writer
-            .event()
-            .key(1)
-            .value(EVENT_VALUE_1)
-            .done()
-            .event()
-            .key(2)
-            .value(EVENT_VALUE_2)
-            .done()
-            .tryWrite();
+        write(
+            w ->
+                w.event()
+                    .key(1)
+                    .value(EVENT_VALUE_1)
+                    .done()
+                    .event()
+                    .key(2)
+                    .value(EVENT_VALUE_2)
+                    .done());
 
     // then
     assertThat(position).isGreaterThan(0);
@@ -146,16 +153,16 @@ public class LogStreamBatchWriterTest {
   public void shouldWriteEventWithValueBuffer() {
     // when
     final long position =
-        writer
-            .event()
-            .key(1)
-            .value(EVENT_VALUE_1)
-            .done()
-            .event()
-            .key(2)
-            .value(EVENT_VALUE_2)
-            .done()
-            .tryWrite();
+        write(
+            w ->
+                w.event()
+                    .key(1)
+                    .value(EVENT_VALUE_1)
+                    .done()
+                    .event()
+                    .key(2)
+                    .value(EVENT_VALUE_2)
+                    .done());
 
     // then
     final List<LoggedEvent> events = getWrittenEvents(position);
@@ -167,16 +174,16 @@ public class LogStreamBatchWriterTest {
   public void shouldWriteEventWithValueBufferPartially() {
     // when
     final long position =
-        writer
-            .event()
-            .key(1)
-            .value(EVENT_VALUE_1, 1, 2)
-            .done()
-            .event()
-            .key(2)
-            .value(EVENT_VALUE_2, 1, 2)
-            .done()
-            .tryWrite();
+        write(
+            w ->
+                w.event()
+                    .key(1)
+                    .value(EVENT_VALUE_1, 1, 2)
+                    .done()
+                    .event()
+                    .key(2)
+                    .value(EVENT_VALUE_2, 1, 2)
+                    .done());
 
     // then
     final List<LoggedEvent> events = getWrittenEvents(position);
@@ -188,16 +195,16 @@ public class LogStreamBatchWriterTest {
   public void shouldWriteEventWithValueWriter() {
     // when
     final long position =
-        writer
-            .event()
-            .key(1)
-            .valueWriter(new DirectBufferWriter().wrap(EVENT_VALUE_1))
-            .done()
-            .event()
-            .key(2)
-            .valueWriter(new DirectBufferWriter().wrap(EVENT_VALUE_2))
-            .done()
-            .tryWrite();
+        write(
+            w ->
+                w.event()
+                    .key(1)
+                    .valueWriter(new DirectBufferWriter().wrap(EVENT_VALUE_1))
+                    .done()
+                    .event()
+                    .key(2)
+                    .valueWriter(new DirectBufferWriter().wrap(EVENT_VALUE_2))
+                    .done());
 
     // then
     final List<LoggedEvent> events = getWrittenEvents(position);
@@ -209,18 +216,18 @@ public class LogStreamBatchWriterTest {
   public void shouldWriteEventWithMetadataBuffer() {
     // when
     final long position =
-        writer
-            .event()
-            .key(1)
-            .value(EVENT_VALUE_1)
-            .metadata(EVENT_METADATA_1)
-            .done()
-            .event()
-            .key(2)
-            .value(EVENT_VALUE_2)
-            .metadata(EVENT_METADATA_2)
-            .done()
-            .tryWrite();
+        write(
+            w ->
+                w.event()
+                    .key(1)
+                    .value(EVENT_VALUE_1)
+                    .metadata(EVENT_METADATA_1)
+                    .done()
+                    .event()
+                    .key(2)
+                    .value(EVENT_VALUE_2)
+                    .metadata(EVENT_METADATA_2)
+                    .done());
 
     // then
     final List<LoggedEvent> events = getWrittenEvents(position);
@@ -232,19 +239,18 @@ public class LogStreamBatchWriterTest {
   public void shouldWriteEventWithMetadataBufferPartially() {
     // when
     final long position =
-        writer
-            .event()
-            .key(1)
-            .value(EVENT_VALUE_1)
-            .metadata(EVENT_METADATA_1, 1, 2)
-            .done()
-            .event()
-            .key(2)
-            .value(EVENT_VALUE_2)
-            .metadata(EVENT_METADATA_2, 1, 2)
-            .done()
-            .tryWrite();
-
+        write(
+            w ->
+                w.event()
+                    .key(1)
+                    .value(EVENT_VALUE_1)
+                    .metadata(EVENT_METADATA_1, 1, 2)
+                    .done()
+                    .event()
+                    .key(2)
+                    .value(EVENT_VALUE_2)
+                    .metadata(EVENT_METADATA_2, 1, 2)
+                    .done());
     // then
     final List<LoggedEvent> events = getWrittenEvents(position);
     assertThat(getMetadataBuffer(events.get(0)))
@@ -257,18 +263,18 @@ public class LogStreamBatchWriterTest {
   public void shouldWriteEventWithMetadataWriter() {
     // when
     final long position =
-        writer
-            .event()
-            .key(1)
-            .value(EVENT_VALUE_1)
-            .metadataWriter(new DirectBufferWriter().wrap(EVENT_METADATA_1))
-            .done()
-            .event()
-            .key(2)
-            .value(EVENT_VALUE_2)
-            .metadataWriter(new DirectBufferWriter().wrap(EVENT_METADATA_2))
-            .done()
-            .tryWrite();
+        write(
+            w ->
+                w.event()
+                    .key(1)
+                    .value(EVENT_VALUE_1)
+                    .metadataWriter(new DirectBufferWriter().wrap(EVENT_METADATA_1))
+                    .done()
+                    .event()
+                    .key(2)
+                    .value(EVENT_VALUE_2)
+                    .metadataWriter(new DirectBufferWriter().wrap(EVENT_METADATA_2))
+                    .done());
 
     // then
     final List<LoggedEvent> events = getWrittenEvents(position);
@@ -280,16 +286,16 @@ public class LogStreamBatchWriterTest {
   public void shouldWriteEventWithKey() {
     // when
     final long position =
-        writer
-            .event()
-            .key(123L)
-            .value(EVENT_VALUE_1)
-            .done()
-            .event()
-            .key(456L)
-            .value(EVENT_VALUE_2)
-            .done()
-            .tryWrite();
+        write(
+            w ->
+                w.event()
+                    .key(123L)
+                    .value(EVENT_VALUE_1)
+                    .done()
+                    .event()
+                    .key(456L)
+                    .value(EVENT_VALUE_2)
+                    .done());
 
     // then
     assertThat(getWrittenEvents(position))
@@ -301,17 +307,17 @@ public class LogStreamBatchWriterTest {
   public void shouldWriteEventWithSourceEvent() {
     // when
     final long position =
-        writer
-            .sourceRecordPosition(123L)
-            .event()
-            .key(1)
-            .value(EVENT_VALUE_1)
-            .done()
-            .event()
-            .key(2)
-            .value(EVENT_VALUE_2)
-            .done()
-            .tryWrite();
+        write(
+            w ->
+                w.sourceRecordPosition(123L)
+                    .event()
+                    .key(1)
+                    .value(EVENT_VALUE_1)
+                    .done()
+                    .event()
+                    .key(2)
+                    .value(EVENT_VALUE_2)
+                    .done());
 
     // then
     final List<LoggedEvent> events = getWrittenEvents(position);
@@ -324,16 +330,16 @@ public class LogStreamBatchWriterTest {
   public void shouldWriteEventWithoutSourceEvent() {
     // when
     final long position =
-        writer
-            .event()
-            .key(1)
-            .value(EVENT_VALUE_1)
-            .done()
-            .event()
-            .key(2)
-            .value(EVENT_VALUE_2)
-            .done()
-            .tryWrite();
+        write(
+            w ->
+                w.event()
+                    .key(1)
+                    .value(EVENT_VALUE_1)
+                    .done()
+                    .event()
+                    .key(2)
+                    .value(EVENT_VALUE_2)
+                    .done());
 
     // then
     final List<LoggedEvent> events = getWrittenEvents(position);
@@ -352,16 +358,16 @@ public class LogStreamBatchWriterTest {
     final ActorFuture<Long> position =
         writerScheduler.call(
             () ->
-                writer
-                    .event()
-                    .key(1)
-                    .value(EVENT_VALUE_1)
-                    .done()
-                    .event()
-                    .key(2)
-                    .value(EVENT_VALUE_2)
-                    .done()
-                    .tryWrite());
+                write(
+                    w ->
+                        w.event()
+                            .key(1)
+                            .value(EVENT_VALUE_1)
+                            .done()
+                            .event()
+                            .key(2)
+                            .value(EVENT_VALUE_2)
+                            .done()));
     writerScheduler.workUntilDone();
 
     // then
@@ -374,15 +380,15 @@ public class LogStreamBatchWriterTest {
   public void shouldNotFailToWriteEventWithoutKey() {
     // when
     final long position =
-        writer
-            .event()
-            .keyNull()
-            .value(EVENT_VALUE_1)
-            .done()
-            .event()
-            .value(EVENT_VALUE_2)
-            .done()
-            .tryWrite();
+        write(
+            w ->
+                w.event()
+                    .keyNull()
+                    .value(EVENT_VALUE_1)
+                    .done()
+                    .event()
+                    .value(EVENT_VALUE_2)
+                    .done());
 
     // then
     assertThat(getWrittenEvents(position)).extracting(LoggedEvent::getKey).contains(-1L);

--- a/service-container/src/main/java/io/zeebe/servicecontainer/CompositeServiceBuilder.java
+++ b/service-container/src/main/java/io/zeebe/servicecontainer/CompositeServiceBuilder.java
@@ -20,7 +20,6 @@ import java.util.Map.Entry;
  * Can be used to install a list of services "transactionally" (in the sense that if one of the
  * services fails to install, all services are removed again).
  */
-@SuppressWarnings({"unchecked", "rawtypes"})
 public class CompositeServiceBuilder {
   private final ServiceContainerImpl container;
   private final ServiceName<Void> compositeServiceName;
@@ -31,7 +30,7 @@ public class CompositeServiceBuilder {
       ServiceName<Void> name,
       ServiceContainer container,
       ServiceName<?>... additionalDependencies) {
-    this.compositeServiceName = name;
+    compositeServiceName = name;
     this.container = (ServiceContainerImpl) container;
 
     final ServiceBuilder<Void> installServiceBuilder =
@@ -71,7 +70,7 @@ public class CompositeServiceBuilder {
   }
 
   class ComposedServiceBuilder<S> extends ServiceBuilder<S> {
-    private ServiceName<S> serviceName;
+    private final ServiceName<S> serviceName;
 
     ComposedServiceBuilder(
         ServiceName<S> name, Service<S> service, ServiceContainerImpl serviceContainer) {
@@ -132,7 +131,10 @@ public class CompositeServiceBuilder {
                     });
               } else {
                 // check all complete
-                if (!future.isDone() && allFutures.stream().allMatch(f -> f.isDone())) {
+                if (!future.isDone()
+                    && allFutures.stream()
+                        .allMatch(f -> f.isDone() && !f.isCompletedExceptionally())) {
+
                   future.completeWith(
                       (CompletableActorFuture<S>) installFutures.get(returnedServiceName));
                   actor.close();

